pkg/controller/util.go

```diff
@@ -109,45 +112,168 @@ func HandleOwnedResourceNamespacedFunc[T metav1.Object](
 	}
 }
 
-// QueuingEventHandler is an implementation of cache.ResourceEventHandler that
+// QueuingEventHandler returns a cache.ResourceEventHandler that
+// simply queues objects that are added/updated/deleted. It skips
+// update events in case the resource version has not changed.
+func QueuingEventHandler(
+	queue workqueue.TypedRateLimitingInterface[types.NamespacedName],
+) cache.ResourceEventHandler {
+	return filteredEventHandler[metav1.Object]{
+		handler: queuingEventHandler{queue: queue},
+		predicates: []predicate.TypedPredicate[metav1.Object]{
+			// prevent unnecessary reconciliations in case the resource did not update
+			predicate.TypedResourceVersionChangedPredicate[metav1.Object]{},
+		},
+	}
+}
+
+// FilterEventHandler returns a cache.ResourceEventHandler that
+// skips events based on the passed predicates and passes all other
+// events to the provided handler.
+func FilterEventHandler[T metav1.Object](
+	handler cache.ResourceEventHandler,
+	predicates ...predicate.TypedPredicate[T],
+) cache.ResourceEventHandler {
+	return filteredEventHandler[T]{
+		handler:    handler,
+		predicates: predicates,
+	}
+}
+
+// queuingEventHandler is an implementation of cache.ResourceEventHandler that
 // simply queues objects that are added/updated/deleted.
-type QueuingEventHandler struct {
-	Queue workqueue.TypedRateLimitingInterface[types.NamespacedName]
+type queuingEventHandler struct {
+	queue workqueue.TypedRateLimitingInterface[types.NamespacedName]
 }
 
+var _ cache.ResourceEventHandler = queuingEventHandler{}
+
 // Enqueue adds a key for an object to the workqueue.
-func (q *QueuingEventHandler) Enqueue(obj interface{}) {
+func (q queuingEventHandler) Enqueue(obj interface{}) {
 	objectName, err := cache.DeletionHandlingObjectToName(obj)
 	if err != nil {
 		runtime.HandleError(err)
 		return
 	}
-	q.Queue.Add(types.NamespacedName{
+	q.queue.Add(types.NamespacedName{
 		Name:      objectName.Name,
 		Namespace: objectName.Namespace,
 	})
 }
 
 // OnAdd adds a newly created object to the workqueue.
-func (q *QueuingEventHandler) OnAdd(obj interface{}, isInInitialList bool) {
+func (q queuingEventHandler) OnAdd(obj interface{}, isInInitialList bool) {
 	q.Enqueue(obj)
 }
 
 // OnUpdate adds an updated object to the workqueue.
-func (q *QueuingEventHandler) OnUpdate(oldObj, newObj interface{}) {
-	if reflect.DeepEqual(oldObj, newObj) {
```

Comment:
This `DeepEqual` is now replaced with the `predicate.TypedResourceVersionChangedPredicate` predicate.